DistroLauncher/Ubuntu/InitTasks.cpp

```diff
@@ -0,0 +1,427 @@
+#include <stdafx.h>
+#include "InitTasks.h"
+
+#include <algorithm>
+#include <charconv>
+#include <exception>
+#include <filesystem>
+#include <fstream>
+#include <sstream>
+#include <utility>
+#include <vector>
+#include <system_error>
+
+namespace Ubuntu {
+
+namespace {
+// Blocks the current thread until all initialization tasks finish.
+void waitForInitTasks(WslApiLoader& api);
+
+// Enforces the existence of a default WSL user either:
+// - defined in /etc/wsl.conf (which might not be in effect yet)
+// - defined in WSL API/registry
+// - or the lowest UID >= 1000 in the NSS passwd database
+// Returns false if a default user couldn't be set.
+bool enforceDefaultUser(WslApiLoader& api);
+}  // namespace
+
+bool CheckInitTasks(WslApiLoader& api, bool checkDefaultUser) {
+  waitForInitTasks(api);
+
+  if (!checkDefaultUser) {
+    return true;
+  }
+
+  return enforceDefaultUser(api);
+}
+
+namespace {
+void waitForInitTasks(WslApiLoader& api) {
+  _putws(L"Checking for initialization tasks...\n");
+
+  DWORD exitCode_ = -1;
+  if (auto hr = api.WslLaunchInteractive(L"which -s cloud-init", FALSE, &exitCode_); FAILED(hr)) {
+    Helpers::PrintErrorMessage(hr);
+    return;
+  }
+  if (exitCode_ != 0) {
+    _putws(L"CheckInitTasks: this release doesn't support initialization tasks.\n");
+    return;
+  }
+
+  exitCode_ = -1;
+  if (auto hr = api.WslLaunchInteractive(L"cloud-init status --wait", FALSE, &exitCode_);
+      FAILED(hr)) {
+    Helpers::PrintErrorMessage(hr);
+    return;
+  }
+  // 0 for "success" or 2 for "recoverable error" - sometimes we get that while status shows "done".
+  // https://cloudinit.readthedocs.io/en/latest/explanation/failure_states.html#cloud-init-error-codes
+  // If that's the case we should inspect the system to check whether we should skip or proceed with
+  // user creation.
+  if (exitCode_ != 0 && exitCode_ != 2) {
+    wprintf(L"CheckInitTasks: failed with exit code: %u\n", exitCode_);
+  }
+}
+
+// Groups the pieces of information from a single user entry in the passwd database we care about.
+struct UserEntry {
+  std::string name;
+  ULONG uid = -1;
+};
+
+namespace fs = std::filesystem;
+fs::path wslConfPath() {
+  // init-once, lazily.
+  static fs::path etcWslConf =
+      fs::path{L"\\\\wsl.localhost"} / DistributionInfo::Name / "etc\\wsl.conf";
```

Comment:
As long as the first argument is a `std::filesystem::path` object the `operator/` performs a concatenation of path segments using the native path separator, being the segments  `std::filesystem::path` or string-like things that are convertible to  `std::filesystem::path`. One extra care we need to take when doing this kind of concatenation is that the right hand side elements must not contain a root segment, otherwise the concatenation becomes replacement.

`"C:\\foo" / "/bar"` => `C:\\foo\\bar` but `"C:\\foo" / "D:\\bar"` => `D:\\bar` on Windows.

That behavior doesn't affect us because `DistributionInfo::Name` and `"etc\\wsl.conf"` won't contain root elements.